slatedb/src/compactor_state.rs

```diff
@@ -78,18 +154,92 @@ impl Display for Compaction {
+        let srs_by_id: HashMap<u32, &SortedRun> =
+            db_state.compacted.iter().map(|sr| (sr.id, sr)).collect();
+
+        sources
+            .iter()
+            .filter_map(|s| s.maybe_unwrap_sorted_run())
+            .filter_map(|id| srs_by_id.get(&id).map(|t| (*t).clone()))
+            .collect()
+    }
+
+    pub(crate) fn get_ssts(db_state: &CoreDbState, sources: &[SourceId]) -> Vec<SsTableHandle> {
+        let ssts_by_id: HashMap<Ulid, &SsTableHandle> = db_state
+            .l0
+            .iter()
+            .map(|sst| (sst.id.unwrap_compacted_id(), sst))
+            .collect();
+
+        sources
+            .iter()
+            .filter_map(|s| s.maybe_unwrap_sst())
+            .filter_map(|ulid| ssts_by_id.get(&ulid).map(|t| (*t).clone()))
+            .collect()
+    }
 }
 
+/// Compaction state persisted in the object store (future work).
+///
+/// This struct will hold the durable snapshot of compaction plans (queued,
+/// in-progress, and completed) keyed by the canonical plan id. Until
+/// persistence is implemented, this is used in tests and as a placeholder for
+/// wiring to the on-disk format.
+#[derive(Clone)]
+pub(crate) struct CompactionState {
+    #[allow(dead_code)]
+    pub(crate) compactor_epoch: u64,
+    // active_compaction plan queued, in-progress and completed mapped by compaction id in object store
+    #[allow(dead_code)]
+    pub(crate) compaction_plans: HashMap<Ulid, CompactionPlan>,
+}
+
+impl CompactionState {
+    #[allow(dead_code)]
+    pub(crate) fn initial() -> Self {
+        Self {
+            compactor_epoch: 0,
+            compaction_plans: HashMap::<Ulid, CompactionPlan>::new(),
+        }
+    }
+}
+
+/// Process-local runtime state owned by the compactor.
+///
+/// This is the in-memory controller view that a single compactor task uses to:
+/// - keep a fresh `DirtyManifest` (view of `CoreDbState`),
+/// - track canonical `compaction_plans` by plan id (ULID), and
+/// - track `scheduled_compactions` by job id for executions owned by this process.
+///
+/// It validates submissions, records lifecycle transitions (Submitted → Pending →
+/// InProgress → Completed/Failed) and mutates the in-memory manifest when jobs
+/// finish.
+///
+/// Difference vs `CompactionState`:
+/// - `CompactorState` is transient, process-local runtime state; it is not
+///   persisted and is rebuilt when the process starts.
+/// - `CompactionState` is the durable, object-store representation (future work)
+///   of compaction plans and their statuses across processes, used for
+///   recovery/GC and history.
 pub struct CompactorState {
     manifest: DirtyManifest,
-    compactions: HashMap<Uuid, Compaction>,
+    // TODO: Add compaction_state during compaction state persistence implementation
+    //compaction_state: DirtyRecord<CompactionState>,
+    compaction_plans: HashMap<Ulid, CompactionPlan>,
```

Comment:
I'm guessing this goes away when compaction_state is used (since it's in compaction_state as well)?